bindings/python/pyiree/test/tree_utils_test.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+from typing import Any
+
+import pyiree as iree
+import pyiree.test
+
+from absl import logging
+from absl.testing import absltest
+import numpy as np
+
+try:
+  import jax
+  JAX_IS_AVALIABLE = True
+except ModuleNotFoundError:
+  JAX_IS_AVALIABLE = False
+
+# pytype thinks pyiree.jax is jax.
+# pytype: disable=module-attr
+
+INPUT_TREE = {
+    "z": [1, (2,), {
+        "y": 3,
+        "x": {
+            "w": (None, 3.14, -46)
+        }
+    }],
+    "v": {
+        "u": 99,
+        "t": np.array([100])
+    }
+}
+LEAVES = [1, 2, 3, None, 3.14, -46, 99, np.array([100])]
+LEAVES_SORTED = [np.array([100]), 99, 1, 2, None, 3.14, -46, 3]
+
+
+class TreeUtilsTest(absltest.TestCase):
+
+  def _assert_ordered_equal(self, first: Any, second: Any):
+    if isinstance(first, dict):
+      self._assert_ordered_equal(list(first.keys()), list(second.keys()))
+      self._assert_ordered_equal(list(first.values()), list(second.values()))
+    else:
+      self.assertEqual(first, second)
+
+  def test_inverse(self):
+    input_tree = copy.deepcopy(INPUT_TREE)
+    output_tree = iree.test.tree_unflatten(
+        *iree.test.tree_flatten(input_tree, sort_keys=False))
+    self._assert_ordered_equal(input_tree, output_tree)
+
+  def test_not_inverse(self):
+    input_tree = copy.deepcopy(INPUT_TREE)
+    output_tree = iree.test.tree_unflatten(*iree.test.tree_flatten(input_tree))
+    self.assertEqual(input_tree, output_tree)
+    with self.assertRaises(AssertionError):
+      self._assert_ordered_equal(input_tree, output_tree)
+
+  def test_leaves_unsorted(self):
+    input_tree = copy.deepcopy(INPUT_TREE)
+    leaves, _ = iree.test.tree_flatten(input_tree, sort_keys=False)
+    self.assertEqual(LEAVES, leaves)
+
+  def test_leaves_sorted(self):
+    input_tree = copy.deepcopy(INPUT_TREE)
+    leaves, _ = iree.test.tree_flatten(input_tree)
+    self.assertEqual(LEAVES_SORTED, leaves)
+
+  def test_not_in_place(self):
+    input_tree = copy.deepcopy(INPUT_TREE)
+
+    leaves, tree_def = iree.test.tree_flatten(input_tree)
+    leaves[0][0] = -100
+    self.assertNotEqual(input_tree["v"]["t"], leaves[0])
```

Comment:
What? Where do "v" and "t" come from? Again this is a case of us having to know something about the magic constant above. Resist the urge to be too dry in unit tests. Unlike in production code, where we're trying to handle all possible cases that might come up, unit tests are about creating minimal input/output pairs that are obviously correct by inspection. I would much rather see `INPUT_TREE` inlined here (and it needn't be the same one as used in other test cases). In places where it just needs to be some input tree, then abstracting it into a constant is great. Arguably this is the case for round-tripping tests as you have above, although even there it's important that it's sufficiently complex to be interesting. But here where the specific structure of it is important to even understand what the tests is doing, just put it in the test.